Fix support for RSA CA keys in Connect.

banks opened the pull request:
Some history is required here:

 - We _theoretically_ added support for RSA keys in Consul CAs in 1.6.0 via https://github.com/hashicorp/consul/pull/6055
 - We missed a few important things though which were initially fixed in https://github.com/hashicorp/consul/pull/6440 but then a couple of issues there and change of owner led to this new branch with a cleaner implementation and more tests

## Design Decisions

Rather than allowing Connect to generate leaf keys that match the CA key type, I've instead kept all leaf keys as EC P256 but allowed for all current CAs to support signing CSR with a different key type.

It's not especially hard to keep key types following CA type and we can change to that approach later but this should work since TLS1.2 relaxed TLS1.1's requirement to use ECDSA when signing a EC certificate. Both Consul and Vault CAs allow signing an EC cert with an RSA key and all of Golang, openssl and Envoy validate these chains correctly (all three are verified by tests in this PR).

For now this seems marginally nicer than having a legacy CA root that needs to be RSA 2048 for compatibility reasons hold back Connect from using more modern crypto. But if we do hit compat issues in the future we can always change to following key types or even allow separate config of CA and leaf key types.

In this current setup any new CA provider will need to be explicitly tested with both key types for the ability to sign EC leaf CSRs. They also need to be able to sign CA CSRs from secondary DCs with a full cross-product. Typically secondary DCs will be configured to use the same CA key type we assume but they may be configured differently which means that all types of CA root need to be able to sign all types of CA CSR and have this tested (see TODO below).

## Testing

I've tried to add a lot of tests to validate the various permutations that can now happen. 

The cert gen tests that use openssl to independently verify our chains are now table-powered to verify more permutations of key type etc.

Both CA providers now have tests that cover them signing a leaf with an EC key with both EC and RSA CA keys. All tests also validate the whole chain afterwards too.

The "unit" tests should be sufficient to cover all behaviour including secondary CA init, but I added an end-to-end integration test using Envoy as a sanity check and in the process found a bug that I'd missed testing at unit test level: the secondary DC CSR would use the wrong signature algorithm when configured with an RSA key. That is fixed and a unit test that would have caught that case was added to by making the test a table test with different key type configs as input.

All the tests also sanity check that the root ca generated actually does have the key type configured too because it was very easy to misconfigure or not catch a config/initialization bug that meant we actually ignored the key type.

Finally, I fixed the issue that we were only populating `CARoots.PrivateKey{Type|Bits}` with the ca config values which were almost always empty but could actually be different from the CA key if it was provided externally. Now we populate that exclusively from looking at the actual key in the CA cert we are adding so it should always be correct. I added table tests to verify some of this parsing and found a few other issues I'll deal with separately (like our use if uint64 for serial which will truncate longer serials from external providers and make them meaningless).

## TODO

 - [x] Document the CA root key fields added in 1.6.0 but not mentioned in docs
 - [x] Document the CA root key config added in 1.6.0 but not added to docs
 - [x] Add explicit unit tests for both CA providers that they can sign a CA CSR with different key type just in case secondaries are configured with a different key type to primaries.

rboyer commented on agent/connect/csr.go:
```diff
@@ -4,19 +4,34 @@ import (
 	"net/url"
 )
 
+// SigAlfoForKey returns the preferred x509.SignatureAlgorithm for a given key
```
```suggestion
// SigAlgoForKey returns the preferred x509.SignatureAlgorithm for a given key
```

rboyer commented on agent/connect/testing_ca.go:
```diff
@@ -36,8 +65,6 @@ func testCA(t testing.T, xc *structs.CARoot, keyType string, keyBits int) *struc
 	signer, keyPEM := testPrivateKey(t, keyType, keyBits)
 	result.SigningKey = keyPEM
 	result.SigningKeyID = EncodeSigningKeyID(testKeyID(t, signer.Public()))
-	result.PrivateKeyType = keyType
```
Is this filled out somewhere else now?

rboyer commented on agent/connect/testing_ca.go:
```diff
@@ -175,7 +202,11 @@ func testLeaf(t testing.T, service string, root *structs.CARoot, keyType string,
 
 	sigAlgo := x509.ECDSAWithSHA256
-	if keyType == "rsa" {
+	rootKeyType, _, err := KeyInfoFromCert(caCert)
```
Maybe it would be a good idea to introduce a `SigAlgoForKeyType` helper like `SigAlgoForKey` that does this similar operation rather than scattering these keytype conditionals around.

banks replied:
Done in next update

rboyer commented on agent/connect/testing_ca.go:
```diff
@@ -218,7 +249,12 @@ func testLeaf(t testing.T, service string, root *structs.CARoot, keyType string,
-	certPEM, keyPEM, err := testLeaf(t, service, root, root.PrivateKeyType, root.PrivateKeyBits)
+	// Currently we only support EC leaf keys and certs even if the CA is using
+	// RSA. We might allow Leafs to follow the signing CA key type later if we
+	// need to for compatibility sake but this is allowe by TLS 1.2 and works with
```
```suggestion
	// need to for compatibility sake but this is allowed by TLS 1.2 and works with
```

rboyer commented on agent/connect/testing_ca_test.go:
```diff
@@ -35,29 +36,33 @@ func testCAAndLeaf(t *testing.T, keyType string, keyBits int) {
-	assert.Nil(err)
-	defer os.RemoveAll(td)
+	require.NoError(err)
+	//defer os.RemoveAll(td)
```
Was this intentional?

banks replied:
nope, thanks

rboyer commented on agent/connect/testing_ca_test.go:
```diff
@@ -35,29 +36,33 @@ func testCAAndLeaf(t *testing.T, keyType string, keyBits int) {
-	defer os.RemoveAll(td)
+	require.NoError(err)
+	//defer os.RemoveAll(td)
+	t.Log(td)
```
Did you meant to keep this logging in the PR?

banks replied:
Nope good catch.

rboyer commented on agent/consul/leader_connect_test.go:
```diff
@@ -24,109 +25,117 @@ import (
-	leafPEM, err := secondaryProvider.Sign(leafCsr)
-	require.NoError(t, err)
+	for _, tt := range tests {
+		t.Run(fmt.Sprintf("%s-%d", tt.keyType, tt.keyBits), func(t *testing.T) {
```
```suggestion
		tt := tt
		t.Run(fmt.Sprintf("%s-%d", tt.keyType, tt.keyBits), func(t *testing.T) {
```
I've gotten into the habit of doing this all the time so that the one time I make the subtest parallel I don't get unexpected test results where all of the subtests use the last testcase configuration.

banks replied:
Does it only matter if you have `t.Parallel()` _inside_ the `t.Run`? Because tons of our suite have `t.Parallel()` at the top level.

I've added these for now although not super sure it's necessary if it only matters if you parallelise the inner test - that seems rare enough that we should maybe not optimize for it?

rboyer replied:
It does only matter if the subtest itself is parallel (Parallel-inside-Run). I originally ran across this surprise in existing ACL tests that were masking a real bug since all but the last case were hidden.

I figured it didn't hurt to dupe the variable to avoid capturing the range variable proactively to avoid future surprises, but I am also not terribly attached to the practice if others think it's just overkill.

For reference the pattern `t.Run(.*\n.*t.Parallel(` occurs 229 times in our tests:
```
$ git grep -A 1 "t\.Run(" | grep -B 1 "t\.Parallel(" | grep -c Run
229
```

banks replied:
Huh I had no idea we do that so often. Yeah I don't think there is harm in the copy as a practice. I wonder how many of those 229 are table tests or some other closure masking failures 🤔.

mkeeler replied:
My 2¢ is that its fine to always capture any table data when executing subtests. Better safe then thinking tests are working when there not.

rboyer commented on agent/consul/leader_connect_test.go:
```diff
@@ -1202,44 +1211,127 @@ func TestLeader_PersistIntermediateCAs(t *testing.T) {
+		{
+			name: "default cert",
+			// Watchout for indentations they will break PEM format
+			pem: `-----BEGIN CERTIFICATE-----
```
Maybe these are bulky enough to warrant stuff them into a `testdata` directory instead of inlining them as code?

rboyer commented on agent/structs/connect_ca.go:
```diff
@@ -102,8 +102,14 @@ type CARoot struct {
-	// Type of private key used to create the CA cert.
+	// PrivateKeyType is the type of the private key used to sign certificates. It
+	// may be "rsa" or "ec". This is provided as a convenience to avoid parsing
+	// the public key from the certificate.
```
```suggestion
	// the private key from the certificate.
```
Did you mean private?

banks replied:
see below

rboyer commented on agent/structs/connect_ca.go:
```diff
@@ -102,8 +102,14 @@ type CARoot struct {
 	PrivateKeyType string
+
+	// PrivateKeyBits is the length of the private key used to sign certificates.
+	// This is provided as a convenience to avoid parsing the public key from the
```
```suggestion
	// This is provided as a convenience to avoid parsing the private key from the
```
Did you mean private?

banks replied:
Actually no - the certificate doesn't contain the private key. This is to help us avoid parsing the certificates _public_ key to infer which type of key it's private key is. I can reword that a little to make it clearer.

banks replied:
Honestly It's a bit of an over-fitted name, I would have preferred `KeyType` since it makes no sense to have public and private keys with different types anyway but I didn't catch that early enough and it would be a breaking change now to fix it.

rboyer commented on agent/structs/connect_ca.go:
```diff
@@ -282,7 +288,18 @@ type CommonCAProviderConfig struct {
 	CSRMaxConcurrent int
 
+	// PrivateKeyType specifies which type of key the CA should generate. It only
+	// applies when the provider is generating it's own key and is ignored if the
```
```suggestion
	// applies when the provider is generating its own key and is ignored if the
```

banks commented:
Realised when adding tests that Vault doesn't actually support cross-signing CA certs with a different key type currently. I filed https://github.com/hashicorp/vault/issues/7709 to see if that's something they might change but it might mean we should for now not bother trying to enforce that full matrix of CA key type compatibility and just document that all DCs need the same CA key type configured. This is likely more future proof for additional CA integrations which may well also not allow signing with different key types too.

I think I'm going to back down on this requirement and instead document the limitation so we can ship this.

banks commented:
@rboyer thanks for the review - I've addressed all the points and finished the TODO items here with some commentary above. Should be ready to go when you get a chance to look!
